examples/sparse/graph_transformer.py

```diff
@@ -0,0 +1,163 @@
+"""
+[A Generalization of Transformer Networks to Graphs]
+(https://arxiv.org/abs/2012.09699)
+"""
+
+import dgl
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from dgl.data import CoraGraphDataset
+from dgl.mock_sparse import bspmm, create_from_coo, mock_bsddmm as bsddmm
+from torch.optim import Adam
+
+
+class SparseMHA(nn.Module):
+    """Sparse Multi-head Attention Module"""
+
+    def __init__(self, hidden_size=80, num_heads=8):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.num_heads = num_heads
+        self.head_dim = hidden_size // num_heads
+        self.scaling = self.head_dim**-0.5
+
+        self.q_proj = nn.Linear(hidden_size, hidden_size)
+        self.k_proj = nn.Linear(hidden_size, hidden_size)
+        self.v_proj = nn.Linear(hidden_size, hidden_size)
+        self.out_proj = nn.Linear(hidden_size, hidden_size)
+
+    def forward(self, A, h):
+        N = len(h)
+        q = self.q_proj(h).reshape(N, self.num_heads, self.head_dim)
+        q *= self.scaling
+        k = self.k_proj(h).reshape(N, self.num_heads, self.head_dim)
+        v = self.v_proj(h).reshape(N, self.num_heads, self.head_dim)
+
+        ######################################################################
+        # (HIGHLIGHT) Compute the multi-head attention with Sparse Matrix API
+        ######################################################################
+        attn = bsddmm(A, q.permute(1, 0, 2), k.permute(1, 2, 0))  # [N, N, nh]
+        attn = attn.softmax()
+        out = bspmm(attn, v.permute(0, 2, 1)).permute(0, 2, 1)
+
+        return self.out_proj(out.reshape(N, -1))
+
+
+class GTLayer(nn.Module):
+    """Graph Transformer Layer"""
+
+    def __init__(self, hidden_size=80, num_heads=8) -> None:
+        super().__init__()
+        self.MHA = SparseMHA(hidden_size=hidden_size, num_heads=num_heads)
+        self.batchnorm1 = nn.BatchNorm1d(hidden_size)
+        self.batchnorm2 = nn.BatchNorm1d(hidden_size)
+        self.FFN1 = nn.Linear(hidden_size, hidden_size * 2)
+        self.FFN2 = nn.Linear(hidden_size * 2, hidden_size)
+
+    def forward(self, A, h):
+        h1 = h
+        h = self.MHA(A, h)
+        h = self.batchnorm1(h + h1)
+
+        h2 = h
+        h = self.FFN2(F.relu(self.FFN1(h)))
+        h = h2 + h
+
+        return self.batchnorm2(h)
+
+
+class GTModel(nn.Module):
+    def __init__(
+        self,
+        in_size,
+        out_size,
+        hidden_size=64,
+        pos_enc_size=2,
```

Comment:
Does this have to be the same as in `laplacian_pe`? If so, consider passing it to both `laplacian_pe` and `GTModel` instead.